File: go/array/leetcode/best_time_buy_sell_stock.go
package main

import (
	"math"
)

//买卖股票的最佳时机I

//暴力法，两层遍历寻找最大值
func maxProfit1(prices []int) int {
	max := 0
	for i := 0; i < len(prices)-1; i++ {
		for j := i + 1; j < len(prices); j++ {
			if prices[j]-prices[i] > max {
				max = prices[j] - prices[i]
			}
		}
	}
	return max
}

//一次遍历
func maxProfit2(prices []int) int {
	min, max := math.MaxInt64, 0
	for _, v := range prices {
		if v < min {
			min = v
		} else if v-min > max {
			max = v - min
		}
	}
	return max
}

File: python/array/leetcode/best_time_buy_sell_stock.py
from typing import List
import sys


# 买卖股票的最佳时机I

class Solution:

    # 暴力法，运行会超时
    def maxProfit_1(self, prices: List[int]) -> int:
        profit = 0
        for i in range(len(prices) - 1):
            for j in range(i + 1, len(prices)):
                profit = max(profit, prices[j] - prices[i])
        return profit


    def maxProfit_2(self, prices: List[int]) -> int:
        profit, minmum = 0, sys.maxsize
        for i in prices:
            if i < minmum:
                minmum = i
            profit = max(profit, i - minmum)
        return profit
        